### DISCUSSION TITLE
CLI with NodeNext module/moduleResolution

### DISCUSSION THREAD

**Discussion by mastercoding** [mikro-orm/mikro-orm#5427] [General]
### Describe the bug

Due to some tsconfig settings being overwritten by the ConfigurationLoader of @mikro-orm/core/utils, the CLI breaks.

My tsconfig.json is currently as follows:

```
{
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    ...
}
```

The module setting is being overwritten in the ConfigurationLoader to "commons", which leaves an incompatible combination of module and moduleResolution ("module": "commons", "moduleResolution": "NodeNext").

Adding "moduleResolution": "classic" to the ConfigurationLoader fixed it for me. I'm not sure if that is the best solution though.

https://github.com/mikro-orm/mikro-orm/blob/01d9e581bdef38c9020563dfd4411080a8cd8a23/packages/core/src/utils/ConfigurationLoader.ts#L147

### Reproduction

See bug

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.1.12

### Node.js version

20.12.1

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> The module setting is being overwritten in the ConfigurationLoader to "commons", which leaves an incompatible combination of module and moduleResolution ("module": "commons", "moduleResolution": "NodeNext").

Its "commonjs", not "commons".

> Adding "moduleResolution": "classic" to the ConfigurationLoader fixed it for me. I'm not sure if that is the best solution though.

~That's is a very new addition to TS, it would break it for people on older versions.~ Actually not, classic is the very old one, not one of the new ones. Still, your suggested change would break it for others. Also, if you check docs, they [explicitly say](https://www.typescriptlang.org/tsconfig#moduleResolution) not to use this option (`'classic' was used in TypeScript before the release of 1.6. classic should not be used.`)

> See bug

What bug? I would very much appreciate a complete reproduction here, I can't fix anything without that, not based on suggestions like this. You haven't shared anything about your setup, instead, you just pointed to a place you would like to see implemented differently.

**Reply by B4nan** [reply]
It works fine for me in the https://github.com/mikro-orm/guide repo, tried to use `nodenext` as in your OP without any differences.

**Reply by B4nan** [reply]
Maybe instead of changing the `module` we should override the `moduleResolution` to `nodenext` too, that makes more sense to me, can you confirm it helps in your case too?

**Reply by vinialbano** [reply]
@B4nan  I am also facing the same problem. I have tried a few things, like using `mikro-orm-esm` with `"type": "module"` in my `package.json` file, but I am still unable to get the CLI to work with TypeScript. I am using MikroORM in a NestJS project, where `module` and `moduleResolution` are already set to `NodeNext` in my `tsconfig.json` file. Additionally, I am running Node.js version `v20.11.1`.

I configured MikroORM in the `package.json` file as requested:
```
"mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts",
      "./dist/mikro-orm.config.js"
    ]
  }
  ```
  
  My `mikro-orm.config.ts` file:
  ```
import { Options } from '@mikro-orm/core';
import { PostgreSqlDriver } from '@mikro-orm/postgresql';
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';

const config: Options = {
  driver: PostgreSqlDriver,
  host: 'localhost',
  port: 5432,
  user: 'postgres',
  password: 'postgres',
  dbName: 'example-database',
  entities: ['dist/**/*.entity.js'],
  entitiesTs: ['src/**/*.entity.ts'],
  metadataProvider: TsMorphMetadataProvider,
  migrations: {
    path: process.cwd() + '/migrations',
  },
  debug: true,
};

export default config;
```

Here's the output of the debug command:
```
$ npm exec mikro-orm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.2.0
   - node 20.11.1
   - typescript 5.4.2
 - package.json found
 - ts-node enabled
 - searched config paths:
   - <path-to-project>/src/mikro-orm.config.ts (found)
   - <path-to-project>/dist/mikro-orm.config.js (not found)
   - <path-to-project>/mikro-orm.config.ts (not found)
   - <path-to-project>/mikro-orm.config.js (not found)
- configuration not found (⨯ Unable to compile TypeScript:
error TS5110: Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.
```

Here's my `tsconfig.json` file:
```
{
    "compilerOptions": {
        "allowJs": true,
        "composite": true,
        "declarationMap": true,
        "esModuleInterop": true,
        "isolatedModules": true,
        "lib": [
            "es2022"
        ],
        "module": "nodenext",
        "moduleDetection": "force",
        "moduleResolution": "nodenext",
        "noUncheckedIndexedAccess": true,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "strict": true,
        "sourceMap": true,
        "target": "es2022",
        "removeComments": true,
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "outDir": "./dist",
        "allowSyntheticDefaultImports": true,
        "resolvePackageJsonExports": true,
        "resolvePackageJsonImports": true,
        "declaration": true,
        "preserveConstEnums": true,
        "incremental": true,
        "useDefineForClassFields": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strictNullChecks": true,
        "strictFunctionTypes": true,
        "strictBindCallApply": true,
        "strictPropertyInitialization": true,
        "alwaysStrict": true,
        "useUnknownInCatchVariables": true
    },
    "include": [
        "src/**/*",
        "tests/**/*"
    ],
    "exclude": [
        "node_modules"
    ]
}
```

**Reply by vinialbano** [reply]
I was now able to make it work by overwriting the `ConfigurationLoader.js` file and setting `moduleResolution` to `nodenext` as you suggested. Pushing a patch should solve it.

**Reply by vinialbano** [reply]
Hi @B4nan. Any updates on this? It seems like an easy fix.
I sent a PR to help.
